File: dotnet/CardsProject/Domain/Marketplace/Order.cs
using CardsProject.Domain.Players;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;

namespace CardsProject.Domain.Marketplace;

public enum OrderStatusType
{
    Pending,
    Paid,
    Processing,
    Shipped,
    Completed,
    Cancelled,
    Refunded
}

public enum OrderPaymentMethodType
{
    Card,
    PayPal,
    Crypto,
    PlatformCredits
}

public class Order : IValidatableObject
{
    public int Id { get; set; }

    public OrderStatusType Status { get; set; }
    public decimal Total { get; set; } = 0.00m;
    public decimal DiscountApplied { get; set; } = 0.00m;
    public string Currency { get; set; } = "USD";
    public OrderPaymentMethodType? PaymentMethod { get; set; }
    public string? PaymentReference { get; set; }
    public string? ShippingAddress { get; set; }
    public string? TrackingNumber { get; set; }
    public DateTime? CreatedAt { get; set; } = null;
    public DateTime? PaidAt { get; set; } = null;
    public DateTime? ShippedAt { get; set; } = null;

    public int? PlayerId { get; set; }
    [ForeignKey(nameof(PlayerId))]
    public Player? Player { get; set; }
    public int? CouponId { get; set; }
    [ForeignKey(nameof(CouponId))]
    public Coupon? Coupon { get; set; }

    // Business operations

    public void Cancel()
    {
        // TODO: implement cancel
    }

    public bool Pay(string paymentRef)
    {
        // TODO: implement pay
        return default;
    }

    public bool ProcessPayment()
    {
        // TODO: implement process_payment
        return default;
    }

    public decimal CalculateTotal()
    {
        // TODO: implement calculate_total
        return default;
    }

    public decimal ApplyDiscount(int percent)
    {
        // TODO: implement apply_discount
        return default;
    }

    public void Refund()
    {
        // TODO: implement refund
    }

    public void NotifyShipped()
    {
        // TODO: implement notify_shipped
    }

    // ── Lifecycle state machine ──────────────────────────────────────
    private static readonly System.Collections.Generic.Dictionary<OrderStatusType, OrderStatusType[]> AllowedTransitions = new()
    {
        [OrderStatusType.Pending] = new[] { OrderStatusType.Paid, OrderStatusType.Cancelled },
        [OrderStatusType.Paid] = new[] { OrderStatusType.Processing, OrderStatusType.Cancelled },
        [OrderStatusType.Processing] = new[] { OrderStatusType.Shipped },
        [OrderStatusType.Shipped] = new[] { OrderStatusType.Completed },
        [OrderStatusType.Completed] = new[] { OrderStatusType.Refunded }
    };

    public void AssertTransition(OrderStatusType to)
    {
        if (!AllowedTransitions.TryGetValue(Status, out var allowed) || !System.Array.Exists(allowed, s => s == to))
            throw new InvalidOperationException($"Transition {Status} -> {to} not allowed");
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( Total >= 0m ))
            yield return new ValidationResult("Order total must not be negative", new[] { nameof(Id) });
        if (!( DiscountApplied <= Total ))
            yield return new ValidationResult("Discount applied cannot exceed order total", new[] { nameof(Id) });
    }
}

File: phoenix/lib/cards_project/marketplace/order.ex
defmodule CardsProject.Marketplace.Order do
  use Ecto.Schema
  import Ecto.Changeset

  schema "orders" do
    field :status, :string
    field :total, :decimal
    field :discount_applied, :decimal
    field :currency, :string, default: "USD"
    field :payment_method, :string
    field :payment_reference, :string
    field :shipping_address, :string
    field :tracking_number, :string
    field :created_at, :naive_datetime
    field :paid_at, :naive_datetime
    field :shipped_at, :naive_datetime
    belongs_to :player, CardsProject.Players.Player
    belongs_to :items, CardsProject.Marketplace.OrderItem
    belongs_to :coupon, CardsProject.Marketplace.Coupon

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:total, :discount_applied, :currency, :created_at, :status, :payment_method, :payment_reference, :shipping_address, :tracking_number, :paid_at, :shipped_at, :player_id, :coupon_id])
    |> validate_required([:total, :discount_applied, :currency, :created_at])
    |> validate_inclusion(:status, ["Pending", "Paid", "Processing", "Shipped", "Completed", "Cancelled", "Refunded"])
    |> validate_inclusion(:payment_method, ["Card", "PayPal", "Crypto", "PlatformCredits"])
    |> validate_number(:total, greater_than_or_equal_to: 0, message: "Order total must not be negative")
    |> then(fn cs ->
      lv = get_field(cs, :total)
      fv = get_field(cs, :discount_applied)
      if not is_nil(lv) and not is_nil(fv) and not (fv <= lv) do
        Ecto.Changeset.add_error(cs, :discount_applied, "Discount applied cannot exceed order total")
      else
        cs
      end
    end)
    |> then(fn cs ->
      if get_field(cs, :status) == "Paid" and (is_nil(get_field(cs, :paid_at))) do
        Ecto.Changeset.add_error(cs, :paid_at, "Paid order must have paid_at set")
      else
        cs
      end
    end)
    |> then(fn cs ->
      if get_field(cs, :status) == "Shipped" and (is_nil(get_field(cs, :tracking_number))) do
        Ecto.Changeset.add_error(cs, :tracking_number, "Shipped order must have a tracking number")
      else
        cs
      end
    end)
    |> then(fn cs ->
      if not is_nil(get_field(cs, :shipped_at)) and (get_field(cs, :status) != "Shipped") do
        Ecto.Changeset.add_error(cs, :status, "shipped_at_requires_shipped_status validation failed")
      else
        cs
      end
    end)
  end

  # ── Business operations ────────────────────────────────────────────

  def cancel(_record) do
    # TODO: implement Order.cancel
    :ok
  end

  def pay(_record, _payment_ref) do
    # TODO: implement Order.pay
    {:error, :not_implemented}
  end

  def process_payment(_record) do
    # TODO: implement Order.process_payment
    {:error, :not_implemented}
  end

  def calculate_total(_record) do
    # TODO: implement Order.calculate_total
    {:error, :not_implemented}
  end

  def apply_discount(_record, _percent) do
    # TODO: implement Order.apply_discount
    {:error, :not_implemented}
  end

  def refund(_record) do
    # TODO: implement Order.refund
    :ok
  end

  def notify_shipped(_record) do
    # TODO: implement Order.notify_shipped
    :ok
  end

  # ── Lifecycle state machine ─────────────────────────────────────────
  @allowed_transitions %{
    "Pending" => ["Paid", "Cancelled"],
    "Paid" => ["Processing", "Cancelled"],
    "Processing" => ["Shipped"],
    "Shipped" => ["Completed"],
    "Completed" => ["Refunded"]
  }

  def assert_transition(%__MODULE__{status: current}, to) do
    allowed = Map.get(@allowed_transitions, current, [])
    if to in allowed do
      :ok
    else
      {:error, "Transition #{current} -> #{to} not allowed"}
    end
  end
end
